### PR TITLE
[WIP] HA Support using NSX-T Load balancers

### PR DESCRIPTION
**Pull request by yastij**
**What this PR does / why we need it**: This is a WIP that would bring support of NSX-T LBs for HA CAPV

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes ##506

**Special notes for your reviewer**:

/assign @akutz 

This isn't tested and not ready for review yet. @akutz is creating the infrastructure to enable e2e for NSX-T

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
add NSX-T as a load balancer provider for CAPV
```

### CODE REVIEW

File: api/v1alpha3/nsxtloadbalancer_types.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	// resources associated with an NSXTLoadBalancer before removing
+	// it from the API server.
+	NSXTLoadBalancerFinalizer = "nsxtloadbalancer.infrastructure.cluster.x-k8s.io"
+)
+
+// NSXTLoadBalancerSpec defines the desired state of NSXTLoadBalancer.
+type NSXTLoadBalancerSpec struct {
+	// lbServiceID is the ID of the NSX-T LoadBalancer Service where virtual servers
+	// for Service Type=LoadBalancer are created
+	LoadBalancerServiceID string `json:"loadBalancerServiceID"`
```
Comment (akutz) [inline api/v1alpha3/nsxtloadbalancer_types.go]: Hi @yastij,

Please correct the GoDoc comment on this field. The first word should match the name of the field.

File: api/v1alpha3/nsxtloadbalancer_types.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+	// vipPoolID is the ID of the IP Pool where VIPs will be allocated
```
Comment (akutz) [inline api/v1alpha3/nsxtloadbalancer_types.go]: Hi @yastij,

Please correct the GoDoc comment on this field. The first word should match the name of the field.

File: controllers/loadbalancer/nsxtloadbalancer_controller.go [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+	releaseAllocatedVIP = false
+
+	return reconcile.Result{}, nil
+}
+
+// controlPlaneMachineToNSXTLoadBalancer is a handler.ToRequestsFunc to be
+// used to trigger reconcile events for an NSXTLoadBalancer when a CAPI
+// Machine is reconciled and it has IP addresses and is a member of the same
+// control plane that the NSXTLoadBalancer services.
+func (r nsxtlbReconciler) controlPlaneMachineToNSXToadBalancer(o handler.MapObject) []ctrl.Request {
```
Comment (akutz) [inline controllers/loadbalancer/nsxtloadbalancer_controller.go]: Hi @yastij,

I bet you can create a common function that can be re-used by this LB and HAProxy that follows a pattern:

```go
ControlPlaneMachineToLoadBalancer(loadBalancerGVK GroupVersionKind) func(handler.MapObject) []ctrl.Request {
  // returns the actual mapping function, using the "loadBalancerGVK" to ensure any loadBalancerRef matches the intended type
}
```

File: pkg/manager/options.go [hunk 4]
```diff
@@ -136,4 +144,12 @@ func (o *Options) defaults() {
 	if o.Password == "" {
 		o.Password = os.Getenv("VSPHERE_PASSWORD")
 	}
+
+	if o.NSXTUsername == "" {
+		o.Username = os.Getenv("NSXT_USERNAME")
```
Comment (akutz) [inline pkg/manager/options.go]: Hi @yastij,

I think this should be `o.NSXTUsername`

File: pkg/manager/options.go [hunk 5]
```diff
@@ -136,4 +144,12 @@ func (o *Options) defaults() {
...
+	}
+
+	if o.NSXTPassword == "" {
+		o.Password = os.Getenv("NSXT_PASSWORD")
```
Comment (akutz) [inline pkg/manager/options.go]: Hi @yastij,

I think this should be `o.NSXTPassword`

**Comment by yastij** [comment]
closing this, as this stalled. This is being reworked right now

**Pull request closed without merging**